README.md

```diff
@@ -1,9 +1,166 @@
+- They configure a mock service (the one here or a custom one) to expect the data defined in the test_data.operation part of the test data file as well as the data to send if that operation is received and what to send if it is not.
+- They stimulate their client under test to send the operation message to the mock service.
+- The mock service will check to see if the operation message it received matches the expected value (this is an exact byte match).
+  - If the match was successful, then it will return the configured match response
+  - Otherwise, it will return the configured non-match response
+- The client under test will attempt to decode the response.  The results can be checked by the test code.
+- If the spec.response.parse_should_succeed field is set to true, then the test code *should* expect a successful parse of the response data.  Otherwise, it *should* expect the parsing to fail.
+
+
+## Writing Tests
+To write a new test, a developer should create a test spec file in the generator/testspecs folder, using the [format below](#test-spec-format). 
+
+They then need to write a generator function in ```generator/generator_lib.py```.  This generator function takes no arguments, and should output a tuple of python binary strings.  
+
+The first element of the tuple should be the protocol buffer encoded value of the operation message that is described in the spec.operation.body_description field of the test spec.  The second element of the tuple should be the protocol buffer encoded value of the result message that is described in the spec.result.body_description field of the test spec.
+
+An example generator for a list opcodes test is shown below:
+
+```python
+def gen_list_opcodes_auth_direct():
+    op = protobuf.list_opcodes_pb2.Operation()
+    op.provider_id = 1
+
+    result = protobuf.list_opcodes_pb2.Result()
+    result.opcodes.extend([1,3,2])
+    return (op.SerializeToString(), result.SerializeToString())
+```
+The generated protocol buffers python library files are in generator/protobuf.
+
+Finally, the developer should add an entry to the generator dictionary at the bottom of the ```generator/generator_lib.py``` file.  The key must correspond to the value in the test spec's spec.generator field.  The value in the generator dictionary must be the new generator function name.  e.g.:
+
+```python
+generators = {
+    'ping_no_auth': gen_ping_no_auth,
+    'list_opcodes_auth_direct': gen_list_opcodes_auth_direct,
+}
+```
+**NOTE**  A generator function can be used in multiple test specs, if required.
+
+## Test Spec Format
+
+Test specs are defined as YAML.  An example file is shown below:
+
+```yaml
+# Copyright 2021 Contributors to the Parsec project.
+# SPDX-License-Identifier: Apache-2.0
+spec:
+  name: list_opcodes_auth_direct
+  generator: list_opcodes_auth_direct
+  description: List opcodes using direct authentication
+  operation:
```

Comment:
I think this should be [`request`](https://parallaxsecond.github.io/parsec-book/parsec_client/wire_protocol.html#requests-and-responses) and the reply `result` - "operation" tends to mean the abstract calls that you can make, whereas the raw encoded data that goes over the wire is a "request".

Reply:
started with that, then of course the protobuf type is an operation - don't mind, can change back to request - i agree, would be clearer

Reply:
done